### ISSUE TITLE
Elasticsearch vector store

### ISSUE THREAD

**Issue by dkarlovi** [hwchase17/langchainjs#771]
Please consider porting the vector store support for ES / OS:
https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/elastic_vector_search.py

# Sidenote

I was playing with using current Opensearch adapter with ES to prototype this, it seems there's a bit more work than expected. Namely, ES v7 doesn't support the `knn` query type OS supports, it only allows [`cosineSimilarity()`](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/query-dsl-script-score-query.html#vector-functions) like implemented in the Python version linked above.

[ES v8 does support `knn`](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#approximate-knn)  (which they call "Approximate kNN") which is supposedly better performing with a large number of documents in ES, but comes with a caveat [the vector dimensions are only up to 1024 (with `index: true` setting which `knn` query type requires)](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html#dense-vector-params) , while [OpenAI embeddings by `text-embedding-ada-002` are by default 1536](https://platform.openai.com/docs/guides/embeddings/second-generation-models).

TLDR:
1. base approach taken in Python version works
2. with a large number of documents and/or advanced features, the new approach should be taken which will cause some implementation issues and tradeoffs.

Update: [with ES 8.8, it now supports 2048 dimensions for vectors](https://github.com/elastic/elasticsearch/pull/95257), meaning ADA embeddings should fit into dense vectors on which ES knows how to do kNN queries.

**Comment by dkarlovi** [comment]
Opensearch in #792

**Comment by zhengyuan-ehps** [comment]
Not sure if this related, but MongoDB Atlas search is also hit the 1024 dimensions restriction when using knn vector field, which means the OpenAi embedding is not compatible with mongo vector store neither.

**Comment by dkarlovi** [comment]
@peterkarman1 you'll not be able to use [the `knn` query](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#approximate-knn), which might not be an issue for you.

Overall, the adapter should be very similar to the existing OpenSearch one, with very minor tweaks, I'm sure PRs are welcome.

**Comment by dkarlovi** [comment]
Closed in #1810.
